byterun/win32.c

```diff
@@ -713,3 +724,113 @@ int caml_snprintf(char * buf, size_t size, const char * format, ...)
   return len;
 }
 #endif
+
+/* Windows Unicode support */
+
+uintnat caml_windows_unicode_runtime_enabled = 1;
+
+static caml_stat_block convert_to_utf16_noexc(UINT codepage, DWORD dwFlags, const char* s)
+{
+    wchar_t* out;
+    int outlen;
+
+    outlen = MultiByteToWideChar(codepage, dwFlags, s, -1, NULL, 0);
+
+    if (outlen == 0)
+      return NULL;
+
+    out = caml_stat_alloc(outlen*sizeof(wchar_t));
+
+    outlen = MultiByteToWideChar(codepage, dwFlags, s, -1, out, outlen);
+
+    CAMLassert (outlen != 0);
+
+    return out;
+}
+
+static caml_stat_block convert_to_utf16(UINT codepage, DWORD dwFlags, const char *s)
+{
+  caml_stat_block b = convert_to_utf16_noexc(codepage, dwFlags, s);
+
+  if (b == NULL)
+    caml_win32_sys_error(GetLastError ());
+
+  return b;
+}
+
+static int convert_from_utf16_length(UINT codepage, DWORD dwFlags, const wchar_t* s)
+{
+  int outlen = WideCharToMultiByte(codepage, dwFlags, s, -1, NULL, 0, NULL, NULL);
+  if (outlen == 0)
+    caml_win32_sys_error(GetLastError ());
+
+  return outlen;
+}
+
+static void convert_from_utf16(UINT codepage, DWORD dwFlags, const wchar_t* s, char *out, int outlen)
+{
+  int res = WideCharToMultiByte(codepage, dwFlags, s, -1, out, outlen, NULL, NULL);
+  CAMLassert (res != 0);
+}
+
+#ifndef WC_ERR_INVALID_CHARS /* For MinGW */
+#define WC_ERR_INVALID_CHARS 0x80
+#endif
+
+CAMLexport value caml_copy_string_of_utf16(const wchar_t *s)
+{
+  UINT codepage;
+  DWORD dwFlags;
+  int outlen;
+  value v;
+
+  if (caml_windows_unicode_runtime_enabled) {
+    codepage = CP_UTF8;
+    dwFlags = WC_ERR_INVALID_CHARS;
+  } else {
+    codepage = CP_THREAD_ACP;
+    dwFlags = 0;
+  }
```

Comment:
On reflection, I agree - I think what would be better would be to have `codepage` and `dwFlags` as (differently named) global variables instead of the single flag?